internal/app/apid/main.go

```diff
@@ -0,0 +1,127 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package main
+
+import (
+	"flag"
+	"log"
+	stdlibnet "net"
+	"os"
+	"strings"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/talos-systems/talos/api/api"
+	"github.com/talos-systems/talos/pkg/config"
+	"github.com/talos-systems/talos/pkg/constants"
+	"github.com/talos-systems/talos/pkg/grpc/factory"
+	"github.com/talos-systems/talos/pkg/grpc/tls"
+	"github.com/talos-systems/talos/pkg/net"
+	"github.com/talos-systems/talos/pkg/startup"
+)
+
+var (
+	configPath *string
+	endpoints  *string
+)
+
+func init() {
+	log.SetFlags(log.Lshortfile | log.Ldate | log.Lmicroseconds | log.Ltime)
+
+	configPath = flag.String("config", "", "the path to the config")
+	endpoints = flag.String("endpoints", "", "the IPs of the control plane nodes")
+
+	flag.Parse()
+}
+
+func main() {
+	if err := startup.RandSeed(); err != nil {
+		log.Fatalf("failed to seed RNG: %v", err)
+	}
+
+	provider, err := createProvider()
+	if err != nil {
+		log.Fatalf("failed to create remote certificate provider: %+v", err)
+	}
+
+	ca, err := provider.GetCA()
+	if err != nil {
+		log.Fatalf("failed to get root CA: %+v", err)
+	}
+
+	tlsConfig, err := tls.New(
+		tls.WithClientAuthType(tls.Mutual),
+		tls.WithCACertPEM(ca),
+		tls.WithCertificateProvider(provider),
+	)
+	if err != nil {
+		log.Fatalf("failed to create OS-level TLS configuration: %v", err)
+	}
+
+	machineClient, err := api.NewLocalMachineClient()
+	if err != nil {
+		log.Fatalf("machine client: %v", err)
+	}
+
+	osClient, err := api.NewLocalOSClient()
+	if err != nil {
+		log.Fatalf("networkd client: %v", err)
+	}
+
+	interceptorProvider, err := createProvider()
```

Comment:
nit: `interceptorProvider` is a bad name IMO. I thought it was a new implementation. In OSD we just called it `provider`.

Reply:
we have a provider for the grpc server itself and then a secondary provider that we pass through to the unaryInterceptor ( `interceptorProvider` ). the provider we pass through to the interceptor is what gets used for auth on the proxied connections

Reply:
Is there not a way to use one for both? We should be able to as we just use it to get the CA and get the certificate. I think I'd rather have this one daemon have one identity.

Reply:
We _could_ use one for both, but it seemed worthwhile to split up the certificates used for apid as a service that clients interact with and certificates used for apid proxy/relay communication. ultimately the certificates are probably just about the same so i'm not sure if there is really any benefit to doing so. seemed more like the right approach.

Reply:
Since they are essentially the same cert, and the idea is that the service should have a single "identity" I think we should go with one cert provider.